### PR TITLE
Send registered metrics to StatsD via StatsDReporter

### PR DESCRIPTION
**Pull request by jordanw-bq**
- Passed smoke test on Qubole cluster
- Currently verifying that metrics are sent to DogStatsD on Qubole cluster using default config settings (localhost:8125)

### CODE REVIEW

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java [hunk 1]
```diff
@@ -113,8 +116,14 @@ public static void startServer(Configuration conf, MetricRegistry metricsRegistr
   /**
    * Register desired metrics.
    */
-  private static void registerMetrics()
+  private static void registerMetrics(Configuration conf)
   {
+    if (CacheConfig.isOnMaster(conf)) {
```
Comment (jordanw-bq) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: @vrajat Added config flag to report to StatsD only from master node

**Comment by vrajat** [comment]
Can you confirm that you have tested on a live cluster ? I looked at the dependencies and they seemed benign.

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java (hunk 1 above)
Comment (vrajat) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: We need one more config to also control whether statsd runs on the master as well. So the if condition needs to be:

    if (CacheConfig.isOnMaster(conf) && CacheConfig.startStatsDReporter(conf)) {

Actually I suggest the following to be future proof:

    if ((CacheConfig.isOnMaster(conf) && CacheConfig.statsdReportOnMaster(conf)) 
       || (!CacheConfig.isOnMaster(conf) && CacheConfig.statsdReportOnWorker(conf))) {

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java [hunk 2]
```diff
@@ -113,8 +116,14 @@ public static void startServer(Configuration conf, MetricRegistry metricsRegistr
...
+      StatsDReporter.forRegistry(metrics)
```
Comment (vrajat) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: Can you assert that values are not null ? What will happen if bad values are given ? Does it show up in logs ?

**Comment by jordanw-bq** [comment]
@vrajat These changes have been tested on a live cluster.

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java (hunk 1 above)
Comment (jordanw-bq) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: What should the default values for these flags be? My thoughts:
```
.isOnMaster -> false (1 master vs. multiple workers)
.reportOnMaster -> true
.reportOnWorker -> false
```

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java (hunk 2 above)
Comment (jordanw-bq) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: Which values? For `metrics` and `conf`?

Comment (vrajat) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: if the value of host or port is wrong.

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java (hunk 1 above)
Comment (vrajat) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: `reportOnMaster` should also be `false`. We will roll it out gradually through QDS cluster configuration.

Comment (jordanw-bq) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: - Flags added to CacheConfig
- Unit tests added for verifying metrics are reported/not reported based on flag combinations
- Flag combinations tested on Qubole cluster

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java (hunk 2 above)
Comment (jordanw-bq) [reply rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeperServer.java]: These values are access through the configuration, so there will always be default values. If the host/port is invalid, StatsDReporter will log an error.

File: rubix-bookkeeper/pom.xml [hunk 3]
```diff
@@ -112,13 +112,19 @@
                             <artifactSet>
                                 <includes>
                                     <include>com.google.guava:guava</include>
+                                    <include>com.readytalk:metrics-statsd-common</include>
+                                    <include>com.readytalk:metrics3-statsd</include>
```
Comment (jordanw-bq) [inline rubix-bookkeeper/pom.xml]: @vrajat StatsDReporter dependencies are now being shaded. 
This has also been tested on a Qubole cluster.

**Pull request merged**